utilities/backup/backup_engine.cc

```diff
@@ -2703,8 +2857,394 @@ void BackupEngineImpl::LoopRateLimitRequestHelper(
   }
 }
 
-void BackupEngineImpl::DeleteChildren(const std::string& dir,
-                                      uint32_t file_type_filter) const {
+IOStatus BackupEngineImpl::InferDBFilesToRetainInRestore(
+    const std::vector<std::pair<const BackupEngineImpl*, const FileInfo*>>&
+        restore_file_infos,
+    const std::unordered_map<uint64_t, std::string>& unowned_sst_backups,
+    const std::string& db_dir, RestoreOptions::Mode mode,
+    std::unordered_set<uint64_t>& files_to_keep) const {
+  if (mode == RestoreOptions::Mode::kPurgeAllFiles) {
+    return IOStatus::OK();
+  }
+  RateLimiter* rate_limiter = options_.restore_rate_limiter.get();
+
+  ROCKS_LOG_INFO(options_.info_log,
+                 "Starting incremental restore evaluation in %" PRIu32 " mode",
+                 mode);
+
+  std::vector<std::string> children;
+  db_fs_->GetChildren(db_dir, io_options_, &children, nullptr)
+      .PermitUncheckedError();  // ignore errors
+
+  // Create a mapping from SST / blob files numbers to their filename and type.
+  std::unordered_map<uint64_t, std::pair<std::string, FileType>>
+      file_num_to_filename_and_type;
+  for (const auto& f : children) {
+    uint64_t number;
+    FileType type;
+    bool ok = ParseFileName(f, &number, &type);
+    if (!ok) {
+      // Couldn't parse existing file name. We deliberately choose to sliently
+      // skip here to avoid noisy & excessive logging in user controlled envs.
+      continue;
+    }
+
+    if (type != kTableFile && type != kBlobFile) {
+      // We only care to optimize restore for large files - like SSTs / blobs.
+      continue;
+    }
+
+    if (type == kBlobFile && mode != RestoreOptions::Mode::kVerifyChecksum) {
+      // Blob files are only supported in kVerifyChecksum mode.
+      continue;
+    }
+
+    file_num_to_filename_and_type[number] = std::make_pair(f, type);
+  }
+
+  // Go through all the unowned SST backups first with the intent to bail out
+  // early if some of them do not have an 'equivalent' db file in db directory.
+  //
+  // NOTE: This is an edge case when restore in `kKeepLatestDbSessionIdFiles`
+  //       mode interleaves with _peculiar_ case of 'excluded files' feature
+  //       where supplied list of (alternative) backup directories are jointly
+  //       missing one or more SST file. In the logic below we try to fill the
+  //       gap by retaining existing db files if db_session_id and sizes match.
+  if (unowned_sst_backups.size() > 0) {
```

Comment:
Can we avoid this extra block of code while handling this unusual case? Perhaps worth chatting about